A40-csds-support.md

```diff
@@ -0,0 +1,328 @@
+
+To improve the debuggability of gRPC, gRPC needs to populate **the reason** and
+**affected resources** when rejecting an update response. This information
+requires the ADS parser to continue parsing past the first error. Here are the
+expected behavior under scenarios:
+
+* If the entire message won’t parse: no need to record anything in CSDS, since
+  we don’t know what type of xDS config or what resources are being updated;
+* If one resource won’t parse: don’t abort the parsing, record the error in the
+  `error_state.details` field. The error string will be attached to other
+  resources that are successfully recognized;
+* If one resource has validation error: record the error, and attach the error
+  string to all resources in the ADS update including itself.
+
+Here is an example of NACK information handling with new ADS changes:
+
+```
+# Imagine we have endpoint A, B, C
+EDS -> {A, B, C}, version 1
+gRPC -> ACK
+CSDS -> [
+  {Endpoint A, ACK, version 1},
+  {Endpoint B, ACK, version 1},
+  {Endpoint C, ACK, version 1}
+]
+
+# The newer endpoint B contains parsing error
+EDS -> {A, B}, version 2
+gRPC -> NACK, Failed to parse endpoint B 
+CSDS -> [
+  {Endpoint A, NACK, version 1, rejected version 2, rejected reason: Failed to parse endpoint B},
+  {Endpoint B, NACK, version 1, rejected version 2, rejected reason: Failed to parse endpoint B},
+  {Endpoint C, ACK, version 1}
+]
+
+# Accepted update will clean error states
+EDS -> {B, C} version 3
+gRPC -> ACK
+CSDS -> [
+  {Endpoint A, NACK, version 1, rejected version 2, rejected reason: Failed to parse endpoint B},
+  {Endpoint B, ACK, version 3},
+  {Endpoint C, ACK, version 3}
+]
+```
+
+
+### xDS Config Error State
+
+When an ADS response is rejected, gRPC should provide debug information via
+CSDS. This is done via `UpdateFailureState` within the `config_dump.proto`,
+which includes the version, timestamp, and the reason of the rejected update
+(see [proto
+definition](https://github.com/envoyproxy/envoy/blob/a4024a578b3f2611fe26229f5d0de99eb0c56895/api/envoy/admin/v3/config_dump.proto#L72)):
+
+```proto
+message UpdateFailureState {
+  ...
+
+  // Time of the latest failed update attempt.
+  google.protobuf.Timestamp last_update_attempt = 2;
+
+  // Details about the last failed update attempt.
+  string details = 3;
+
+  // This is the version of the rejected resource.
+  string version_info = 4;
+}
+```
+
+### CSDS Service Design
+
+Ideally, the CSDS service should be reusable by our users, instead of just an
+internal tool for the gRPC library. So, the CSDS service should be implemented
+in wrapper languages and Java/Go. There are three major parts in the service
+design:
+
+* Cache resources in the global XdsClient;
+* Collect additional metadata about the cached resource (update timestamp,
+  version, etc.);
+* Assemble cached resources into the gigantic config dump response.
+
+Note that xDS config should include the `config.core.v3.Node` information in the
```

Comment:
I don't really see a compelling reason to do this.  The reason that the node field is present in CSDS is that CSDS was originally designed to be served by control planes, not clients, and the control plane needed a way to indicate which response was for which client.  That's not relevant when you're talking to one individual client, so it's not clear to me that there's any benefit in the client populating this field.

Reply:
Added some context here: https://github.com/grpc/proposal/pull/223#discussion_r583764176

Reply:
The Node information is helpful to debug gRPC and control plane simultaneously. In common cases, users connect to the CSDS endpoint  with ip:port, so the Node information can be new to them.

It also works if a user sets configuration in the control plane, and not observing the expected behavior. Then the user can use this to check what xDS config gRPC receives, and use the Node information to query the control plane's version of the xDS configs. This is how "diff" works in `istioctl` (https://istio.io/latest/docs/ops/diagnostic-tools/proxy-cmd/#retrieve-diffs-between-envoy-and-istiod).

Reply:
As mentioned elsewhere, I'm not arguing that this info might not be useful, just that CSDS is not the right way to expose it.  Exposing this info is a separate problem, and if we want to solve that, we should provide a separate RPC service for it.